File: low-level-design-java/src/inmemorycachedesign/cacheservice/CacheService.java
package inmemorycachedesign.cacheservice;


import inmemorycachedesign.constants.EvictionPolicy;
import inmemorycachedesign.dao.CacheDao;

public class CacheService {
    private static CacheService instance = null;
    private CacheService(){

    }

    public static CacheService getInstance(){
        if(instance == null){
            instance = new CacheService();
        }
        return instance;
    }

    private final CacheDao cacheDao = CacheDao.getInstance();

    public void initialize(int size, EvictionPolicy evictionPolicy){
        cacheDao.createCache(size, evictionPolicy);
    }

    public Object get(Object object){
        return cacheDao.get(object);
    }

    public void put(Object key, Object value){
        cacheDao.put(key, value);
    }

    public Boolean delete(Object key){
        return cacheDao.delete(key);
    }

    public Boolean clear(){
        return cacheDao.clear();
    }

    public void display(){
        cacheDao.display();
    }


}

File: low-level-design-python/inmemorycachedesign/cacheservice/CacheService.py
from inmemorycachedesign.dao.CacheDao import CacheDao


class CacheService:
    __instance = None

    def __init__(self):
        self.cache_dao = None

    def __new__(cls):
        if cls.__instance is None:
            cls.__instance = super(CacheService, cls).__new__(cls)
        return cls.__instance

    def initialize(self, size, eviction_policy):
        self.cache_dao = CacheDao()
        self.cache_dao.create_cache(size, eviction_policy)

    def get(self, key):
        return self.cache_dao.get(key)

    def put(self, key, value):
        self.cache_dao.put(key, value)

    def delete(self, key):
        return self.cache_dao.delete(key)

    def clear(self):
        return self.cache_dao.clear()

    def display(self):
        self.cache_dao.display()
